elkscmd/man/man1/mkfs.1
.TH MKFS 1
.SH NAME
mkfs \- make a file system
.SH SYNOPSIS
\fBmkfs \fR[\fB\-Ldot\fR] [\fB\-B \fIblocksize\fR] [\fB\-i \fIinodes\fR] [\fB\-b \fIblocks\fR] \fIspecial \fIprototype\fR
.br
.SH OPTIONS
.TP 5
.B \-L
# Make a listing on standard output
.TP 5
.B \-d
# Use mod time of \fImkfs\fR binary for all files
.TP 5
.B \-o
# Use a drive other than 0 or 1 (safety precaution)
.TP 5
.B \-t
# Do not test if file system fits on the medium
.TP 5
.B \-1
# Make a version 1 file system (for backward compatibility)
.TP 5
.B \-i
# Number of i-nodes (files)
.TP 5
.B \-B
# Filesystem block size (in bytes)
.TP 5
.B \-b
# Filesystem size (in blocks)
.TP 5
.B \-x
# Extra space after dynamic sizing (blocks and inodes)
.SH EXAMPLES
.TP 20
.B mkfs /dev/fd1 proto
# Make a file system on \fI/dev/fd1\fR
.TP 20
.B mkfs -b 360 /dev/fd1
# Make empty 360 block file system
.TP 20
.B mkfs /dev/fd1 360
# Alternate way to specify the size
.SH DESCRIPTION
.PP
.I Mkfs
builds a file system and copies specified files to it.
The prototype file tells which directories and files to copy to it.
If the prototype file cannot be opened, and its name is just a string of
digits, an empty file system will be made with the specified number of
blocks.
A sample prototype file follows.
The text following the \fI#\fR sign in the example below is comment.
In real prototype files, comments are not allowed.
.PP
.nf
.ta 0.20i 0.70i 1.10i 3i 3.5i 4i
	boot			# boot block file (ignored)
	360 63			# blocks and i-nodes
	d--755 1 1		# root directory
	   bin	d--755 \|2 1	# bin dir: mode (755), uid (2), gid (1)
		sh	\|---755 2 1 /user/bin/shell	# shell has mode \fIrwxr-xr-x\fP
		mv	-u-755 2 1 /user/bin/mv	# u = SETUID bit
		login	-ug755 2 1 /user/bin/login	# SETUID and SETGID
	   $			# end of \fI/bin\fP
	   dev	d--755 2 1	# special files: tty (char), fd0 (block)
		tty	c--777 2 1 4 0	# uid=2, gid=1, major=4, minor=0
		fd0	b--644 2 1 2 0 360	# uid, gid, major, minor, blocks
	   $			# end of \fI/dev\fP
	   user	d--755 12 1	# user dir: mode (755), uid (12), gid (1)
		ast	d--755 12 1	# \fI/user/ast\fP
		$		# \fI/user/ast\fP is empty
	   $			# end of \fI/user\fP
	$			# end of root directory
.PP
.fi
The first entry on each line (except the first 3 and the $ lines, which
terminate directories) is the name the file or directory will get on the
new file system.  
Next comes its mode, with the first character being
\fB\-dbcs\fR for regular files, directories, block special files and character 
special files, symlinks, respectively.
The next two characters are used to specify the SETUID and SETGID bits, as
shown above.
The last three characters of the mode are the 
.I rwx
protection bits.
.PP
Following the mode are the uid and gid.
For special files, the major and minor devices are needed.
.PP
The maximum size of a file system is 1 Gb for a version 2 file system,
and 64 Mb for a version 1 file system.  Alas the 8086
.I fsck
runs out of memory on a V2 file system larger than 128 Mb, so for the 8086
version of
\s-1MINIX 3\s-1
you have to limit yourself to file systems of that size.
.SH "SEE ALSO"
.BR mkproto (1),
.BR fsck (1),
.BR mount (1).


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/disk_utils/mkfs.c
/*
 * mkfs.c - make a linux (minix) file-system.
 *
 * (C) 1991 Linus Torvalds. This file may be redistributed as per
 * the Linux copyright.
 */

/***************************************************************************
 *
 * 24.11.1991 -	time began. Used the fsck sources to get started.
 *
 * 25.11.1991 -	corrected some bugs. Added support for ".badblocks"
 *		The algorithm for ".badblocks" is a bit weird, but
 *		it should work. Oh, well.
 *
 * 25.01.1992 -	Added the -l option for getting the list of bad blocks
 *              out of a named file. (Dave Rivers, rivers@ponds.uucp)
 *
 * 28.02.1992 -	added %-information when using -c.
 *
 * 28.02.1993 -	added support for other namelengths than the original
 *		14 characters so that I can test the new kernel routines..
 *
 * 09.10.1993 -	make exit status conform to that required by fsutil
 *		<faith@cs.unc.edu>
 *
 * 31.10.1993 -	added inode request feature, for backup floppies: use
 *		32 inodes, for a news partition use more.
 *		Scott Heavner <sdh@po.cwru.edu>
 *
 * 03.01.1994 -	Added support for file system valid flag.
 *		Dr. Wettstein <greg%wind.uucp@plains.nodak.edu>
 *
 ***************************************************************************
 *
 * Usage:  mkfs [-c] [-nXX] [-iXX] device size-in-blocks
 *         mkfs [-l filename ] device size-in-blocks
 *
 * The device may be a block device or a image of one, but this isn't
 * enforced (but it's not much fun on a character device :-). 
 *
 */

#include <errno.h>

#include <stdio.h>
#include <time.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <fcntl.h>
/*#include <ctype.h>*/
#include <stdlib.h>
#include <termios.h>
#include <sys/stat.h>

#include <linuxmt/fs.h>
#include <linuxmt/minix_fs.h>

/* Check for gcc removed from here. */

#ifndef __linux__
#define volatile
#endif

#define MINIX_ROOT_INO 1
#define MINIX_BAD_INO 2

#define TEST_BUFFER_BLOCKS 16
#define MAX_GOOD_BLOCKS 512

#define UPPER(size,n) ((size+((long)(n)-1))/(n))
#define INODE_SIZE (sizeof(struct minix_inode))
#define INODE_BLOCKS UPPER(INODES,MINIX_INODES_PER_BLOCK)
#define INODE_BUFFER_SIZE (INODE_BLOCKS * BLOCK_SIZE)

#define BITS_PER_BLOCK (BLOCK_SIZE<<3)

static char * device_name = NULL;
static int DEV = -1;
static long BLOCKS = 0;

static int dirsize = 16;
static int magic = MINIX_SUPER_MAGIC;
static unsigned int ikl;

static char root_block[BLOCK_SIZE];

static char * inode_buffer = NULL;
#define Inode (((struct minix_inode *) inode_buffer)-1)
static char super_block_buffer[BLOCK_SIZE];
#define Super (*(struct minix_super_block *)super_block_buffer)
#define INODES (Super.s_ninodes)
#define ZONES (Super.s_nzones)
#define IMAPS (Super.s_imap_blocks)
#define ZMAPS (Super.s_zmap_blocks)
#define FIRSTZONE (Super.s_firstdatazone)
#define ZONESIZE (Super.s_log_zone_size)
#define MAXSIZE (Super.s_max_size)
#define MAGIC (Super.s_magic)
#define NORM_FIRSTZONE (2+IMAPS+ZMAPS+INODE_BLOCKS)

/* Max of 64k zones and 64k inodes */
static char inode_map[BLOCK_SIZE * 8];
static char zone_map[BLOCK_SIZE * 8];

unsigned char set_bit(unsigned int nr, void * add)
{
	unsigned int * addr = add;
	int	mask, retval;

	addr += nr >> 4;
	mask = 1 << (nr & 0x0f);
	retval = (mask & *addr) != 0;
	*addr |= mask;
	return retval;
}

unsigned char clear_bit(unsigned int nr, void * add)
{
	unsigned int * addr = add;
	int	mask, retval;

	addr += nr >> 4;
	mask = 1 << (nr & 0x0f);
	retval = (mask & *addr) != 0;
	*addr &= ~mask;
	return retval;
}

unsigned char test_bit(unsigned int nr, void * add)
{
	unsigned int * addr = add;
	int	mask;

	addr += nr >> 4;
	mask = 1 << (nr & 0x0f);
	return ((mask & *addr) != 0);
}

#define bit(addr,bit) test_bit(bit,addr)
#define setbit(addr,bit) set_bit(bit,addr)
#define clrbit(addr,bit) clear_bit(bit,addr)

#define inode_in_use(x) (bit(inode_map,(x)))
#define zone_in_use(x) (bit(zone_map,(x)-FIRSTZONE+1))

#define mark_inode(x) (setbit(inode_map,(x)))
#define unmark_inode(x) (clrbit(inode_map,(x)))

#define mark_zone(x) (setbit(zone_map,(x)-FIRSTZONE+1))
#define unmark_zone(x) (clrbit(zone_map,(x)-FIRSTZONE+1))

/*
 * Volatile to let gcc know that this doesn't return. When trying
 * to compile this under minix, volatile gives a warning, as
 * exit() isn't defined as volatile under minix.
 */
volatile void fatal_error(const char * fmt_string,int status)
{
	printf(fmt_string);
	exit(status);
}

#define usage() fatal_error("Usage: mkfs /dev/name blocks (Max blocks=65535)\n",16)
#define die(str) fatal_error("mkfs: " str "\n",8)

void write_tables(void)
{
	Super.s_state |= MINIX_VALID_FS;
	Super.s_state &= ~MINIX_ERROR_FS;

	if (BLOCK_SIZE != lseek(DEV, (long)BLOCK_SIZE, SEEK_SET)) {
		die("seek failed in write_tables");
	}
	if (BLOCK_SIZE != write(DEV, super_block_buffer, BLOCK_SIZE))
		die("unable to write super-block");
	if (IMAPS*BLOCK_SIZE != write(DEV,inode_map,IMAPS*BLOCK_SIZE))
		die("Unable to write inode map");
	if (ZMAPS*BLOCK_SIZE != write(DEV,zone_map,ZMAPS*BLOCK_SIZE))
		die("Unable to write zone map");
	if (BLOCK_SIZE != write(DEV,inode_buffer,BLOCK_SIZE))
		die("Unable to write inodes");
	for (ikl=1;ikl<INODE_BLOCKS;ikl++) {
		if (BLOCK_SIZE != write(DEV,inode_buffer + BLOCK_SIZE,BLOCK_SIZE))
			die("Unable to write inodes");
	}
}
 
void write_block(int blk, char * buffer)
{
	unsigned long seek_point = (long)((long)blk * BLOCK_SIZE);

/*	printf("Seeking in user space to %ld\n",seek_point); */
/*	printf("Seeked to %ld\n",lseek(DEV, seek_point, SEEK_SET)); */
	if (seek_point != lseek(DEV, seek_point, SEEK_SET)) {
		die("seek failed in write_block");
	}
	if (BLOCK_SIZE != write(DEV, buffer, BLOCK_SIZE))
		die("write failed in write_block");
}

int get_free_block(void)
{
	int blk;

	blk = FIRSTZONE;
	while (blk < ZONES && zone_in_use(blk))
		blk++;
	if (blk >= ZONES)
		die("not enough good blocks");
	mark_zone(blk);
	return blk;
}

void make_root_inode(void)
{
	struct minix_inode * inode = &Inode[MINIX_ROOT_INO];

	mark_inode(MINIX_ROOT_INO);
	inode->i_zone[0] = get_free_block();
	inode->i_nlinks = 2;
	inode->i_time = time(NULL);
	root_block[2*dirsize] = '\0';
	root_block[2*dirsize+1] = '\0';
	inode->i_size = 2*dirsize;
	inode->i_mode = S_IFDIR + 0755;
	write_block(inode->i_zone[0],root_block);
}

void setup_tables(void)
{
	unsigned i;

	memset(inode_map,0xff,sizeof(inode_map));
	memset(zone_map,0xff,sizeof(zone_map));
	memset(super_block_buffer,0,BLOCK_SIZE);
	MAGIC = magic;
	ZONESIZE = 0;
	/* volume limit is 7+512+512L*512L but set max_size to blocks in fs*/
	MAXSIZE = BLOCKS*1024L;
	ZONES = BLOCKS;
	INODES = BLOCKS/3;
#if 0 /* FIXME: add ability to specify inode count rather than guess*/
	if ( BLOCKS > 32768L ) INODES += (BLOCKS-32768)*4/3;
	if ( INODES > 63424L ) INODES = 63424L;
	if ((INODES & 8191) > 8188)
		INODES -= 5;
	if ((INODES & 8191) < 10)
		INODES -= 20;
#endif
	if (INODES > 32736)		/* max inodes to fit in 4 imap blocks*/
		INODES = 32736;
	IMAPS = UPPER(INODES,BITS_PER_BLOCK);
	ZMAPS = 0;
	while (ZMAPS != UPPER(BLOCKS - NORM_FIRSTZONE,BITS_PER_BLOCK))
		ZMAPS = UPPER(BLOCKS - NORM_FIRSTZONE,BITS_PER_BLOCK);
	FIRSTZONE = NORM_FIRSTZONE;
	for (i = FIRSTZONE ; i<ZONES ; i++)
		unmark_zone(i);
	for (i = MINIX_ROOT_INO ; i<INODES ; i++)
		unmark_inode(i);
	inode_buffer = malloc(2048);
	if (!inode_buffer)
		die("Unable to allocate buffer for inodes");
	memset(inode_buffer,0,2048);
	printf("%u inodes\n",INODES);
	printf("%u blocks\n",ZONES);
	printf("Firstdatazone=%d (%d)\n",FIRSTZONE,NORM_FIRSTZONE);
	printf("Zonesize=%d\n",BLOCK_SIZE<<ZONESIZE);
	printf("Maxsize=%ld\n\n",MAXSIZE);
}


int main(int argc, char ** argv)
{
	int i;
	char * tmp;
	struct stat statbuf;
	char * listfile = NULL;

	if (INODE_SIZE * MINIX_INODES_PER_BLOCK != BLOCK_SIZE)
		die("bad inode size");

	if ((argc == 3) && (argv[1][0] != '-') && (argv[2][0] != '-')) {
		BLOCKS = strtol(argv[2],&tmp,0);
		if (*tmp) {
			usage();
		}
		device_name = argv[1];
	} else
		usage();
	if (!device_name || BLOCKS<10L || BLOCKS > 65535L) {
		usage();
	}
	tmp = root_block;
	tmp[0] = 1;
	tmp[1] = 0;
	strcpy(tmp+2,".");
	tmp += dirsize;
	tmp[0] = 1;
	tmp[1] = 0;
	strcpy(tmp+2,"..");
	tmp += dirsize;
	tmp[0] = 2;
	tmp[1] = 0;
	strcpy(tmp+2,".badblocks");
	DEV = open(device_name,O_RDWR );
	if (DEV<0)
		die("unable to open device");
	if (fstat(DEV,&statbuf)<0)
		die("unable to stat %s");
	//else if (statbuf.st_rdev == 0x0300 || statbuf.st_rdev == 0x0340)
		//die("Will not try to make filesystem on '%s'");

	setup_tables();
	make_root_inode();
	write_tables();
	sync();
	return 0;
}
